Add a FilterEmptyDatasetsTool and allow filtering tools to apply to arbitrary collection structures

alimatai opened the pull request:
Hello,

Here is an extra-condition for the tool ``FilterFailedDatasetsTool`` to allow it to remove empty datasets as well as failed datasets in a collection. I tested it on my data and it seems to work fine. Since it's my first time looking at galaxy's source code, I don't know if there is a proper way to do that ... And I didn't make the improvement for paired datasets because I don't have proper data to test it. But I assume it would be something like ``if forward_o.is_ok and reverse_o.is_ok and forward_o.has_data() and reverse_o.has_data():`` ?

Commit:
Add a condition to FilterFailedDatasetsTool to remove empty files

Hello,

Here is an extra-condition for the tool ``FilterFailedDatasetsTool`` to allow it to remove empty datasets as well as failed datasets in a collection. I tested it on my data and it seems to work fine. Since it's my first time looking at galaxy's source code, I don't know if there is a proper way to do that ... And I didn't make the improvement for paired datasets because I don't have proper data to test it. But I assume it would be something like ``if forward_o.is_ok and reverse_o.is_ok and forward_o.has_data() and reverse_o.has_data():`` ?

mvdbeek commented:
Hey @Mataivic, this is an interesting improvement, but I think this behaviour should be selectable in the tool form, since empty datasets are usually not failed datasets (if they are failed that should be configured in the tool that produced the output). You can define additional parameters in the xml of this tool (https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/tools/filter_failed_collection.xm), just like you'd do for regular tools. The parameters are then available in `incoming`, which is a simple dictionary, see for example https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/tools/__init__.py#L2567 how to use this.

Commit:
improved filtering of empty files in a collection (1)

- Supplemental parameter : 'remove_empty' which allow the user to choose if he/she wants to keep or not empty files in the collection.
- Coming next : the modified wrapper.
(Edited directly on GitHub).

Commit:
modified wrapper filter_failed_collection

- Supplemental 'select' parameter : 'remove_empty'

alimatai commented:
@mvdbeek It may not be the most elegant way to do, but I applied your suggestion. I also made the improvement for paired dataset but I have no idea if it works the same way than regular collections (I guess the Travis test will tell us that) ; feel free to cancel that one or to modify it.

alimatai commented:
@mvdbeek Hello, Just remembered I made this PR ... Is is acceptable ?

mvdbeek commented:
Sorry, forgot about this, I do have a couple of changes that I'll try to get in.

mvdbeek committed:
Simplify and modularize FilterFailedDatasetsTool class

mvdbeek committed:
Add filter empty collection tool

I think having a separate tool for this task makes sense and makes the
distinction between empty and fialed datasets clearer.

mvdbeek commented:
OK, so this took far longer than I had hoped, sorry for holding this up so long.
Ultimately I think it's better to have a separate tools for filtering failed and empty datasets from a collection, so I've added 2 commits. This should now work for any type of collection structure, not just list, paired or list:paired.

mvdbeek commented:
@galaxybot test this

nsoranzo commented on lib/galaxy/tools/filter_empty_collection.xml:
```diff
@@ -0,0 +1,41 @@
+    </tests>
+    <help><![CDATA[
+
+This tool takes a list dataset collection and filters out empty datasets. This is useful for continuing a multi-sample analysis when downstream tools require datasets to have content.
```
`s/list dataset collection/dataset collection/`

nsoranzo commented on lib/galaxy/tools/__init__.py:
```diff
@@ -2510,42 +2510,64 @@ def produce_outputs(self, trans, out_data, output_collections, incoming, history
-    require_dataset_ok = False
+class FilterDatasetsTool(DatabaseOperationTool):
+
+    def get_new_elements(self, history, elements_to_copy):
```
Maybe make it "private" by renaming it to `_get_new_elements` ?

Also, maybe swap the `history` and `elements_to_copy` parameters in line with how the method is called below?

mvdbeek replied:
I've flipped the caller methods, that seems more "natural" to me and that's how we handle a lot of the signatures in the test framework

nsoranzo commented on lib/galaxy/tools/__init__.py:
```diff
@@ -2510,42 +2510,64 @@ def produce_outputs(self, trans, out_data, output_collections, incoming, history
+                elements_to_copy.append(element)
+            elif collection_type == 'paired':
+                # One of the pairs is not OK, skip creating output for this pair
+                return
```
Is it OK to just `return` here or should this return e.g. `self._outputs_dict()` as `DatabaseOperationTool.produce_outputs()` does?

mvdbeek replied:
Good point, it actually isn't

mvdbeek committed:
Change get_new_elements to _get_new_elements

And use same order for arguments in signature and caller.

mvdbeek committed:
Correct help text of filter_failed_collection and filter_empty_collection tools

mvdbeek committed:
Add tests for filtering list:pair and list:list:pair collections

mvdbeek committed:
Map over list or list:paired collection in filter collection tools

nsoranzo commented on lib/galaxy/tools/__init__.py:
```diff
@@ -2510,42 +2510,68 @@ def produce_outputs(self, trans, out_data, output_collections, incoming, history
+            # A pair
+            elements = collection.collection.elements
+            collection_type = collection.collection.collection_type
+        # We only process list or paired collections. Higher order collection will be mapped over
```
Should lines 2533 and 2536 be updated to say `list of pair` instead of `pair`/`paired collections` ?

mvdbeek committed:
Update comments to the fact that we're processing list or list:paired collections

mvdbeek commented:
Thanks for putting this up here @Mataivic, very useful functionality!

alimatai commented:
@mvdbeek You're welcome :). Thanks for your help !
